server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java

```diff
@@ -266,6 +269,80 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakS
         }
     }
 
+    /**
+     * Creates a new {@link KeycloakSession} and runs the specified callable in a new transaction. If the transaction fails
+     * with a SQL retriable error, the method re-executes the specified callable until it either succeeds or the maximum number
+     * of attempts is reached, leaving some increasing random delay milliseconds between the invocations. It uses the exponential
+     * backoff + jitter algorithm to compute the delay. More details https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+     *
+     * @param factory a reference to the {@link KeycloakSessionFactory}.
+     * @param callable a reference to the {@link KeycloakSessionTaskWithResult} that will be executed in a retriable way.
+     * @param attemptsCount the maximum number of attempts to execute the callable.
+     * @param retryIntervalMillis the base interval value in millis used to compute the delay.
+     * @param <V> the type returned by the callable.
+     * @return the value computed by the callable.
+     */
+    public static <V> V runJobInRetriableTransaction(final KeycloakSessionFactory factory, final KeycloakSessionTaskWithResult<V> callable,
+                                                     final int attemptsCount, final int retryIntervalMillis) {
+        int retryCount = 0;
+        Random rand = new Random();
+        V result;
+        while (true) {
+            KeycloakSession session = factory.create();
+            KeycloakTransaction tx = session.getTransactionManager();
+            try {
+                tx.begin();
+                result = callable.run(session);
+                if (tx.isActive()) {
+                    if (tx.getRollbackOnly()) {
+                        tx.rollback();
+                    } else {
+                        tx.commit();
+                    }
+                }
+                break;
+            } catch (RuntimeException re) {
+                if (tx.isActive()) {
+                    tx.rollback();
+                }
+                if (isExceptionRetriable(re) && ++retryCount < attemptsCount) {
+                    int delay = (int)(Math.pow(2, retryCount) * retryIntervalMillis) + rand.nextInt(retryIntervalMillis);
```

Comment:
Please use shift left operation rather than `Math.pow(2, ..)`, and limit wait time:
```suggestion
                    int delay = Math.min(retryIntervalMillis * attempts, (1 << retryCount) * retryIntervalMillis)) + rand.nextInt(retryIntervalMillis);
```

Reply:
True, the Javadoc should be updated accordingly.

Reply:
Fixed javadoc.

Reply:
@sguilhen @hmlnarik @ahus1 Sorry to bring up this old conversation but I have seen we compute the delay differently in [Retry](https://github.com/keycloak/keycloak/blob/main/common/src/main/java/org/keycloak/common/util/Retry.java#L116) and here which should be based on the same algorithm. In `KeycloakModelUtils` we compute the delay for the current iteration and then add a random number from 0 to `retryIntervalMillis`. However, based on the [algorithm](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/) we should generate a random number between 0 and `Math.min(retryIntervalMillis * attemptsCount, (1 << retryCount)`. Is this on purpose or should I report an issue for this?

Reply:
Ideally we should have the same implementation at all places. However there are multiple variants of the exponential backoff algorithm, and the important aspect is to have a randomness involved, as the link also states. If _exponential backoff and jitter_ algorithm provides better results in terms of throughput, then we should go that way.
At this point, this is a low priority issue, but nice to have.

Reply:
Agree with @hmlnarik that we need to have the same algorithm in all places. I would open an issue to revisit the backoff + jitter algorithm and make it consistent in the different classes.